presto-main/src/main/java/com/facebook/presto/metadata/SessionPropertyManager.java

```diff
@@ -87,95 +85,127 @@ public void addSystemSessionProperties(List<PropertyMetadata<?>> systemSessionPr
     }
 
-    private <T> void addSessionPropertyInternal(Optional<String> catalogName, PropertyMetadata<T> sessionProperty)
+    public Optional<PropertyMetadata<?>> getSystemSessionPropertyMetadata(String name)
     {
-        SessionProperty<T> value = new SessionProperty<>(catalogName, sessionProperty);
-        allSessionProperties.put(value.getFullyQualifiedName(), value);
+        requireNonNull(name, "name is null");
+
+        return Optional.ofNullable(systemSessionProperties.get(name));
     }
 
-    public PropertyMetadata<?> getSessionPropertyMetadata(String name)
+    public Optional<PropertyMetadata<?>> getConnectorSessionPropertyMetadata(ConnectorId connectorId, String propertyName)
     {
-        requireNonNull(name, "name is null");
-
-        SessionProperty<?> sessionProperty = allSessionProperties.get(name);
-        if (sessionProperty == null) {
-            throw new PrestoException(INVALID_SESSION_PROPERTY, "Unknown session property " + name);
+        requireNonNull(connectorId, "connectorId is null");
+        requireNonNull(propertyName, "propertyName is null");
+        Map<String, PropertyMetadata<?>> properties = connectorSessionProperties.get(connectorId);
+        if (properties == null || properties.isEmpty()) {
+            throw new PrestoException(INVALID_SESSION_PROPERTY, "Unknown connector " + connectorId);
         }
-        return sessionProperty.getMetadata();
+
+        return Optional.ofNullable(properties.get(propertyName));
     }
 
-    public List<SessionPropertyValue> getAllSessionProperties(Session session)
+    public List<SessionPropertyValue> getAllSessionProperties(Session session, Map<String, ConnectorId> catalogs)
     {
         requireNonNull(session, "session is null");
 
-        List<SessionPropertyValue> properties = new ArrayList<>();
-        for (SessionProperty<?> sessionProperty : allSessionProperties.values()) {
-            PropertyMetadata<?> propertyMetadata = sessionProperty.getMetadata();
-            String defaultValue = firstNonNull(propertyMetadata.getDefaultValue(), "").toString();
-
-            Map<String, String> values;
-            if (sessionProperty.getCatalogName().isPresent()) {
-                values = session.getCatalogProperties(sessionProperty.getCatalogName().get());
-            }
-            else {
-                values = session.getSystemProperties();
-            }
-            String value = values.getOrDefault(sessionProperty.getPropertyName(), defaultValue);
-
-            properties.add(new SessionPropertyValue(
+        ImmutableList.Builder<SessionPropertyValue> sessionPropertyValues = ImmutableList.builder();
+        Map<String, String> systemProperties = session.getSystemProperties();
+        for (PropertyMetadata<?> property : systemSessionProperties.values()) {
+            String defaultValue = firstNonNull(property.getDefaultValue(), "").toString();
```

Comment:
This can be `nullToEmpty(property.getDefaultValue())`

(the `toString()` is redundant)